Integrate algorithm instantiation, front end, and script interface

jdaymude opened the pull request:
Integrates the logic for registering algorithms with the simulator's script interface and front end sidebar.

Resolves #3.

ryanyiu committed:
~ Replace scriptinterface instantiation with algo class instantiation

~ Edit the Algorithm class to include signals to fire for logging errors and for setting the system
~ Add child classes that inherit from Algorithm to handle algorithm-specific instantiation

jdaymude commented on main/application.cpp:
```diff
@@ -54,6 +54,12 @@ Application::Application(int argc, char *argv[])
+    for (Algorithm* alg : parameterModel->getAlgorithmList()->getAlgs()) {
+      connect(alg, SIGNAL(log), [qmlRoot](const QString msg, const bool isError){
+        QMetaObject::invokeMethod(qmlRoot, "log", Q_ARG(QVariant, msg), Q_ARG(QVariant, isError));
+      });
```
My compile errors all point here. AFAIK, `connect` is supposed to have four parameters: a reference to the slot object, the slot function, a reference to the signal object, and the signal function. You only seem to have three of the four (I think you're missing the signal object?) Also, it's unclear to me if the second parameter should be `SIGNAL(log)` or `SIGNAL(log(QString, bool))` to capture the parameters.

ryanyiu replied:
Sorry about that, I had something different that didn't have compile errors! I tried copying the same method for hooking up the `log` signal with the corresponding slot that we have in line 102 of `application.cpp`, but it gave me build errors regarding `Q_OBJECT`. I just pushed the change to fix the compile errors, so if you try building, you should see the error that I'm talking about

ryanyiu committed:
~Resolve compilation errors

jdaymude commented on ui/algorithm.h:
```diff
@@ -5,13 +5,17 @@
+#include <core/system.h>
+
+class Algorithm : public QObject {
  public:
```
You're missing a `Q_OBJECT` guard right here. See, for example, `core/simulator.h`.

jdaymude replied:
That will at least get the thing to build, but you'll probably need to do a lot more testing to make sure everything actually works.

ryanyiu committed:
+ Enable calls to instantiate functions on simulator-side

ryanyiu committed:
~ Change calls for scripting

ryanyiu committed:
~ Instantiate simulator with ShapeFormation when initialized

- Remove debug statements
~ Updated the case where a path to a script is passed to account for changes

ryanyiu committed:
- Remove testing files

jdaymude commented on AmoebotSim.pro:
```diff
@@ -74,3 +74,5 @@ OTHER_FILES += \
     res/qml/A_ResultTextField.qml \
     res/qml/A_TextField.qml \
     res/qml/main.qml
+
```
Not sure this is needed.

jdaymude commented on core/simulator.h:
```diff
@@ -20,9 +20,9 @@ class Simulator : public QObject {
   Simulator();
   virtual ~Simulator();
 
-  void setSystem(std::shared_ptr<System> _system);
```
None of the changes in this file change anything, and they make the formatting worse, I think... revert?

jdaymude commented on ui/algorithm.cpp:
```diff
@@ -2,6 +2,14 @@
+#include "alg/leaderelection.h"
+#include "alg/shapeformation.h"
+
 #include "ui/algorithm.h"
```
Small formatting thing: `#include "ui/algorithm.h"` should come first, right after the header comment, because it is the header for this source file. See the [style guide](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes).

jdaymude commented on script/scriptengine.cpp:
```diff
@@ -10,13 +10,25 @@
   : scriptInterface(new ScriptInterface(*this, sim, vis)) {
   // Create a global object for the JavaScript engine and make its methods
   // globally accessible. The engine owns the script interface.
+  algList = list;
```
For simple assignment from a constructor parameter to a member variable, you can do this in the initializer list:
```
ScriptEngine::ScriptEngine(Simulator& sim, VisItem* vis, AlgorithmList* list)
    : scriptInterface(new ScriptInterface(*this, sim, vis)),
      _algList(list) {
   ...
}
```

jdaymude commented on script/scriptengine.h:
```diff
@@ -30,6 +32,7 @@ class ScriptEngine : public QObject {
  private:
   QJSEngine engine;
   ScriptInterface* scriptInterface;
+  AlgorithmList* algList;
```
I realize we hadn't been doing this in this file, but member variables should be preceded by an underscore: `_algList`.

ryanyiu replied:
So, did you want me to do it for all of the member variables, like `_engine` and `_scriptInterface`?

jdaymude commented on script/scriptengine.h:
```diff
@@ -18,7 +19,8 @@ class ScriptEngine : public QObject {
  public:
-  ScriptEngine(Simulator& sim, VisItem* vis = nullptr);
+  ScriptEngine(Simulator& sim, VisItem* vis = nullptr,
+               AlgorithmList* list = nullptr);
```
Easiest if parameter names match the variable they're going to be assigned to, so consider using `AlgorithmList* algList` to be assigned to `_algList`.

jdaymude commented on script/scriptengine.cpp:
```diff
@@ -10,13 +10,25 @@
+    // the JS function to instantiate it is
+    // 'shapeformation(numParticles, holeProb, mode)'
+    QString JScmd = "this[" + alg->getSignature() + "] = " + alg->getSignature() + "[instantiate]";
+    engine.evaluate(JScmd);
```
Seems redundant to create the `JScmd` variable and then evaluate it but not use it elsewhere. Could just `engine.evaluate()` the string directly.

jdaymude commented on ui/algorithm.cpp:
```diff
@@ -46,50 +54,165 @@ void Algorithm::addParameter(QString parameter, QString defaultValue) {
+  Algorithm("Basic Shape Formation", "shapeformation") {
+  addParameter("# Particles", "200");
+  addParameter("Hole Prob.", "0.2");
+  addParameter("Shape", "\"h\"");
```
I think wrapping the `h` in quotes is causing a bug. If you run the simulator and then press the instantiate button without specifying parameters in the sidebar, it tries to use `\"h\"` as the default and then gives the error that it isn't one of the supported modes (it's expecting `h` alone).

jdaymude commented on script/scriptengine.cpp:
```diff
@@ -10,13 +10,25 @@
+    // Create a string that will register each algorithm's instantiate function
+    // as a global function under its signature name, e.g., for Shape Formation
+    // the JS function to instantiate it is
+    // 'shapeformation(numParticles, holeProb, mode)'
```
Suggestion for rewriting this comment, just to make it crystal clear (also potentially move it up above the for loop): "For each algorithm, register it with the script engine and associate its signature (e.g., 'shapeformation' for the Basic Shape Formation algorithm) with its ::instantiate() function defined in ui/algorithm.*"

jdaymude commented on ui/algorithm.cpp:
```diff
@@ -46,50 +54,165 @@ void Algorithm::addParameter(QString parameter, QString defaultValue) {
+  addParameter("Hole Prob.", "0.2");
+}
+
+void InfObjCoatingAlg::instantiate(const int numParticles, const double holeProb) {
```
This is just a couple characters over the line width.

jdaymude commented on ui/algorithm.h:
```diff
@@ -30,12 +35,68 @@ class Algorithm {
 
+ signals:
+  void log(const QString msg, bool error = false);
+  void setSystem(std::shared_ptr<System> _system);
```
The prefix underscore should only be used for member variables.

jdaymude commented on ui/algorithm.h:
```diff
@@ -5,13 +5,18 @@
 
 #include <QString>
 #include <QStringList>
+#include <QObject>
```
These should be alphabetized, with `#include <QObject>` coming first.

jdaymude commented on ui/algorithm.h:
```diff
@@ -5,13 +5,18 @@
+#include <QObject>
 
-class Algorithm {
+#include <core/system.h>
```
Just for style, our libraries should be in quotations, not angle brackets.

jdaymude commented on ui/algorithm.h:
```diff
@@ -30,12 +35,68 @@ class Algorithm {
+class DiscoDemoAlg : public Algorithm {
+ public:
+  DiscoDemoAlg();
+
```
No need for the extra space between the constructor and instantiate function in all of these class definitions.

jdaymude commented on ui/parameterlistmodel.cpp:
```diff
@@ -62,20 +62,40 @@ void ParameterListModel::updateAlgParameters(QString algName) {
 void ParameterListModel::createCommand(QString algName) {
```
Given how you've updated this algorithm (i.e., how it does the instantiating now as opposed to just creating a string that the ScriptEngine processes), you may want to consider renaming it and updating its documentation in the `.h` file.

ryanyiu committed:
~ Minor structural and comment fixes

jdaymude committed:
~ Updated DiscoDemo tutorial to match new algorithm registration method


~ Minor formatting improvements

jdaymude committed:
- Removed signal from ParameterListModel that is no longer used

ryanyiu committed:
~ Fix the issue of functions not being registered in scripts

jdaymude committed:
~ Updated DiscoDemo tutorial to match final registration changes

~ Updated GIF at the end of DiscoDemo to match new UI

jdaymude commented:
LGTM.

ryanyiu reviewed (approved):
LGTM
